sql/sem/tree: fix width handling for BIT column default values

spilchen opened the pull request:
Previously, when a column was being added or dropped and had NOT NULL with no default value, the parseComputedExpr logic would generate a placeholder value to avoid NULL violations (see #46285 for historical context). For columns of type BIT(n), this fallback value was incorrectly set as BIT(0), which has a width mismatch and caused errors during row insertions or deletions.

This change ensures that the fallback value matches the column's specified BIT(n) width, allowing operations to proceed without errors during schema changes involving such columns.

Fixes #152326
Fixes https://github.com/cockroachdb/cockroach/issues/149567
Fixes https://github.com/cockroachdb/cockroach/issues/151451

Release note (bug fix): Fixed a bug where INSERTs could fail with a type checking error while adding a BIT(n) column.

Epic: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152910)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/152910#-OZF3d6E-3H1EtphJMLs:-OZF3d6E-3H1EtphJMLt:b-natiri):*
nit: we can note this as a bug fix. e.g. "Fixed a bug where INSERTs could fail with a type checking error while adding a BIT(n) column."

since it can affect foreground app traffic, we might want to backport this.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/sem/tree: fix width handling for BIT column default values

Previously, when a column was being added or dropped and had NOT NULL
with no default value, the parseComputedExpr logic would generate a
placeholder value to avoid NULL violations (see #46285 for historical
context). For columns of type BIT(n), this fallback value was
incorrectly set as BIT(0), which has a width mismatch and caused errors
during row insertions or deletions.

This change ensures that the fallback value matches the column's
specified BIT(n) width, allowing operations to proceed without errors
during schema changes involving such columns.

Fixes #152326

Release note (bug fix): Fixed a bug where INSERTs could fail with a type
checking error while adding a BIT(n) column.

Epic: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152910)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/152910#-OZF3d6E-3H1EtphJMLs:-OZFM_-s9DbgM2_G-mLM:b-wvr8ko):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we can note this as a bug fix. e.g. "Fixed a bug where INSERTs could fail with a type checking error while adding a BIT(n) column."

since it can affect foreground app traffic, we might want to backport this.
</blockquote></details>

Good call on the release note. I can add the backport label too, but maybe address that part when you get back from PTO.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 15 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152910#-:-OZG9sDb2ZHVpIWpOeR5:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152910#-:-OZGKghc-z0iM8ERwk3l:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->
